frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -283,42 +292,44 @@ impl<T: Config> SignedSubmissions<T> {
 	/// to `is_score_better`, we do not change anything.
 	pub fn insert(&mut self, submission: SignedSubmissionOf<T>) -> InsertResult<T> {
 		// verify the expectation that we never reuse an index
-		debug_assert!(!self.indices.values().any(|&idx| idx == self.next_idx));
-
-		let weakest = match self.indices.try_insert(submission.raw_solution.score, self.next_idx) {
-			Ok(Some(prev_idx)) => {
-				// a submission of equal score was already present in the set;
-				// no point editing the actual backing map as we know that the newer solution can't
-				// be better than the old. However, we do need to put the old value back.
-				self.indices
-					.try_insert(submission.raw_solution.score, prev_idx)
-					.expect("didn't change the map size; qed");
-				return InsertResult::NotInserted
-			},
-			Ok(None) => {
-				// successfully inserted into the set; no need to take out weakest member
-				None
-			},
-			Err((insert_score, insert_idx)) => {
-				// could not insert into the set because it is full.
-				// note that we short-circuit return here in case the iteration produces `None`.
-				// If there wasn't a weakest entry to remove, then there must be a capacity of 0,
-				// which means that we can't meaningfully proceed.
-				let weakest_score = match self.indices.iter().next() {
+		debug_assert!(!self.indices.iter().map(|(_, _, x)| x).any(|&idx| idx == self.next_idx));
+		let block_number = frame_system::Pallet::<T>::block_number();
+
+		let weakest = match self.indices.try_push((
+			submission.raw_solution.score,
+			block_number,
+			self.next_idx,
+		)) {
+			Ok(_) => None,
+			Err(_) => {
+				// the queue is full -- if this is better, insert it.
+				let weakest_score = match self.indices.iter().next().defensive() {
 					None => return InsertResult::NotInserted,
-					Some((score, _)) => *score,
+					Some((score, _, _)) => *score,
 				};
 				let threshold = T::BetterSignedThreshold::get();
 
 				// if we haven't improved on the weakest score, don't change anything.
-				if !insert_score.strict_threshold_better(weakest_score, threshold) {
+				if !submission.raw_solution.score.strict_threshold_better(weakest_score, threshold)
+				{
 					return InsertResult::NotInserted
 				}
 
-				self.swap_out_submission(weakest_score, Some((insert_score, insert_idx)))
+				self.swap_out_submission(
+					0, // swap out the worse one, which is always index 0.
+					Some((submission.raw_solution.score, block_number, self.next_idx)),
+				)
 			},
 		};
 
+		// this is the ONLY place that we insert, and we sort post insertion. If scores are the
+		// same, we sort based on reverse of submission block number.
+		self.indices
+			.sort_by(|(score1, bn1, _), (score2, bn2, _)| match score1.cmp(score2) {
+				Ordering::Equal => bn1.cmp(&bn2).reverse(),
+				x @ _ => x,
```

Comment:
should be sufficient

```suggestion
				x => x,
```